lib/adb-device.js

```diff
@@ -190,6 +190,64 @@ class ADBDevice {
     return data;
   }
 
+  // Receive a file listing and parse it into an array of file objects.
+  async recvList(localId, remoteId) {
+    // This is the structure of each directory entry returned by LIST
+    // see https://android.googlesource.com/platform/system/core/+/master/adb/SYNC.TXT#39
+    const props = [
+      'cmd',
+      'mode',
+      'size',
+      'modified',
+      'length',
+      'filename'
+    ];
+    const files = [];
+    let file = {};
+    let i = 0;
+
+    // Keep requesting new packets until we receive 'DONE'
+    // TODO: Do we need a timeout in case 'DONE' is never sent? Shouldn't happen
+    while (true) {
+      let packet = await this.recvAndOkay(localId, remoteId);
+      let pos = 0;
+      do {
```

Comment:
Ahhh I get it now. They do the same thing when sending back a file via `PULL`.
See these:
- recvFile: https://github.com/appium/node-adb-client/blob/master/lib/adb-device.js#L176-L191
- parseFileData: https://github.com/appium/node-adb-client/blob/master/lib/helpers.js#L105-L122

I'm not a huge fan of my own solution either, but there's only so much we can do to work around the data the ADB daemon gives us. :)

Perhaps @jlipps and @imurchie have some ideas.